engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,575 @@ package ingestion
+				s.stopBoundary.StopHeight-1,
+				h.Height,
+			)
+		return
+	}
+
+	s.stopExecution()
 }
 
 func (s *StopControl) stopExecution() {
-	if s.crash {
-		s.log.Fatal().Msgf(
-			"Crashing as finalization reached requested "+
-				"stop height %d and the highest executed block is (%d - 1)",
-			s.stopHeight,
-			s.stopHeight,
-		)
+	log := s.log.With().
+		Stringer("requested_stop", s.stopBoundary).
+		Uint64("last_executed_height", s.stopBoundary.StopHeight).
+		Stringer("last_executed_id", s.stopBoundary.stopAfterExecuting).
+		Logger()
+
+	s.stopped = true
+	log.Warn().Msg("Stopping as finalization reached requested stop")
+
+	if s.stopBoundary != nil && s.stopBoundary.ShouldCrash {
+		// TODO: crash more gracefully or at least in a more explicit way
+		log.Fatal().Msg("Crashing as finalization reached requested stop")
 		return
 	}
+}
 
-	s.log.Debug().
-		Int8("previous_state", int8(s.state)).
-		Int8("new_state", int8(StopControlPaused)).
-		Msg("StopControl state transition")
+func (s *StopControl) handleVersionBeacon(
+	height uint64,
+) error {
+	if s.nodeVersion == nil || s.stopped {
+		return nil
+	}
 
-	s.state = StopControlPaused
+	if s.versionBeacon != nil && s.versionBeacon.SealHeight >= height {
+		// we already processed this or a higher version beacon
+		return nil
+	}
 
-	s.log.Warn().Msgf(
-		"Pausing execution as finalization reached "+
-			"the requested stop height %d",
-		s.stopHeight,
-	)
+	vb, err := s.versionBeacons.Highest(height)
+	if err != nil {
+		return fmt.Errorf("failed to get highest "+
+			"version beacon for stop control: %w", err)
+	}
 
-}
+	if vb == nil {
+		// no version beacon found
+		// this is unexpected as there should always be at least the
+		// starting version beacon, but not fatal.
+		// It can happen if the node starts before bootstrap is finished.
+		s.log.Info().
+			Uint64("height", height).
+			Msg("No version beacon found for stop control")
+		return nil
+	}
 
-// executingBlockHeight should be called while execution of height starts,
-// used for internal tracking of the minimum possible value of stopHeight
-func (s *StopControl) executingBlockHeight(height uint64) {
-	// TODO: should we lock here?
+	if s.versionBeacon != nil && s.versionBeacon.SealHeight >= vb.SealHeight {
+		// we already processed this or a higher version beacon
+		return nil
+	}
 
-	if s.state == StopControlPaused {
-		return
+	s.log.Info().
+		Uint64("vb_seal_height", vb.SealHeight).
+		Uint64("vb_sequence", vb.Sequence).
+		Msg("New version beacon found")
+
+	// this is now the last handled version beacon
+	s.versionBeacon = vb
+
+	// this is a new version beacon check what boundary it sets
+	stopHeight, shouldBeSet, err := s.getVersionBeaconStopHeight(vb)
+	if err != nil {
+		return err
 	}
 
-	// updating the highest executing height, which will be used to reject setting
-	// stopHeight that is too low.
-	if height > s.highestExecutingHeight {
-		s.highestExecutingHeight = height
+	set := s.stopBoundary != nil
```

Comment:
the stop boundary update logic and related logging seems strange.  Instead of clearing existing stop boundary, we should pick the earliest stop boundary?


```
if !shouldBeSet {
  return nil
}

canChange, reason := s.canChangeStop(...)
if !canChange {
  log.Warn() ...
  return nil
}

newStopBoundary := &stopBoundary{...}
if s.stopBoundary == nil || newStopBoundary.LessThan(s.stopBoundary) {
  s.log.Info().
    Stringer("old_stop", s.stopBoundary).
    Stringer("new_stop", newStop).
    Msg("New stop set")

  s.stopBoundary = newStop
}

return nil
```

Reply:
(also see comment in SetStop)

Reply:
the logging was weird... I removed the `log.Info().Msg("stop cleared")` log it was something I forgot to delete.